### ISSUE TITLE
[Feature] Support custom cursors (for cursor chat)

### ISSUE THREAD

**Issue by TodePond** [tldraw/tldraw#1438]
We want to add more fun things to our cursors! Like...

- cursor chat
- customisation

How can we achieve this?

# Let's talk about cursors
On the web, there are three types of cursor.

## 1. browser cursor
You can use one of the built-in named cursors
eg: "pointer", "drag", etc

For the purposes of this issue, I've made my browser cursors **green** to make them stand out more.

✅ Uses your own cursor (OS or custom).
✅ Uses your chosen cursor size (a11y).
✅ Doesn't lag behind your pointer.
❌ Can only choose from a limited set.
❌ Inconsistent - different platforms have different cursors.

![240183445-8e7bacb0-65dc-4d99-ad72-ae03af911197](https://github.com/tldraw/tldraw/assets/15892272/ae9cdf5d-6665-4537-a8ac-9b57275c956d)

## 2. image cursor
You can provide an image to use as the cursor.

❌ Doesn't use your own cursor.
✅ Uses your chosen cursor size (a11y).
❌ But it can look stretched/pixelated for bigger sizes.
✅ Doesn't lag behind your pointer.
✅ We can customise the cursor (eg: rotation, colour).
❌ Limited to 32x32 images for some browsers (but we can provide up to 128x128 for those that support it).
✅ We can keep the cursor the same across platforms.

![240184558-0093d95b-4eb3-4a43-96c0-6d2d5f868626](https://github.com/tldraw/tldraw/assets/15892272/4ddb8eac-9523-4418-9ca8-236159d48df3)

## 3. floating cursor
You can hide the cursor, and float something in its place, like an SVG.

❌ Doesn't use your own cursor.
❌ Doesn't use your chosen cursor size (a11y).
❌ Lags behind your pointer.
✅ We can use *anything* as the cursor.
✅ We can keep the cursor the same across platforms.

![240185863-ba7a6c0b-0ee1-4255-b15d-7251c1fe80ac](https://github.com/tldraw/tldraw/assets/15892272/a5478330-8d89-41a1-a828-a9255bae1799)

# What gets used?
In reality, most apps use a combination of the three approaches.

## tldraw
We use **browser cursors** most of the time.
We use **image cursors** when we need to control rotation.

✅ This keeps things pretty simple! (while allowing us to control rotation in some cases)

![240186555-6d6f04ab-949f-4676-bb76-ace1d9de1342](https://github.com/tldraw/tldraw/assets/15892272/544701de-06f1-4ff1-bba3-6fd7d9e2548f)

It does have some downsides:
❌ We can't customise the **browser cursors** at all (eg: changing their colour).

![240188628-1f245acc-a66f-439f-ae44-5442917284f8](https://github.com/tldraw/tldraw/assets/15892272/eab5f515-2323-4238-b369-e911da522fc3)

❌ We can't do more complex things with the **image cursors** (eg: nametags).

![240189407-1b8dd3cd-5277-4b51-8134-01d43529fc75](https://github.com/tldraw/tldraw/assets/15892272/08a43edc-9339-4965-ac50-1cce579c2410)

❌ Our **image cursors** sometimes look out-of-place compared to the **browser cursors**.
But this is kinda fine. We're not replacing *all* cursors, just the _weird ones_!

![240191272-6ea4f94c-b89f-43ec-80cc-6fa4f9500823](https://github.com/tldraw/tldraw/assets/15892272/c70b5fca-6332-4b58-a762-ed565e3863ec)

## Excalidraw

They use **browser cursors** everywhere (I think).

✅ Nice and simple! Everything's consistent.

![240210424-6b4dd529-2830-4bdb-8485-ad4eba0b5121 (1)](https://github.com/tldraw/tldraw/assets/15892272/508ea749-ef5e-4121-ac47-7da49188e1db)

❌ But when it comes to rotation, cursors point the wrong way slightly.

![240196037-1d5cf11f-bf3a-44d9-b31a-1d97318631d5](https://github.com/tldraw/tldraw/assets/15892272/c95861c7-c2a1-4aa1-ad79-f49679bbdce9)

PS: There's a bug - when you click the handle of a rotated shape, it uses the wrong cursor.

![2023-05-23 at 15 48 52 - Olive Leech](https://github.com/tldraw/tldraw/assets/15892272/a56886ae-243a-461c-a9dd-dd2269d72164)

## [React Cursor Chat](https://commerce-cursor-chat.vercel.app/)

- They use **floating cursors** for everything.

✅ It lets them put anything they want onto the cursor.

![240197146-8133afd0-19d3-40fa-891c-54bfc58423cd](https://github.com/tldraw/tldraw/assets/15892272/e9b9dc5d-f02f-46e2-a131-7dbdf5b631b5)

❌ It always lags behind the pointer.

![240197536-262d8c50-ea1a-411b-b0c9-9cad43fcae8b](https://github.com/tldraw/tldraw/assets/15892272/822605ef-81eb-49d8-ab19-51e342e9c106)

## FigJam

- They use **browser cursors** on the UI.
- They use **image cursors** on the canvas.
- They use **floating cursors** for cursor chat.

When you start cursor-chatting, they _swap_ between **image** and **floating**.
✅ This lets them show more complex stuff.

![240198466-b73795ea-61cf-43a7-9306-78c1d850a8f8](https://github.com/tldraw/tldraw/assets/15892272/2d2b06c4-70f4-47e7-ab8b-b57c9702a4c6)

❌ But it can cause a change in cursor size.

![240199297-b4311463-20ca-48c3-bd71-f40ec09f1607](https://github.com/tldraw/tldraw/assets/15892272/b46efed4-a3b5-43c1-a9ce-404ae0e674d8)

❌ It adds a noticeable lag to the cursor (but only when you're chatting).

![240199712-2abf3e53-661d-433d-9fe4-e30d68f0d23f](https://github.com/tldraw/tldraw/assets/15892272/dfec3fc6-6653-4634-b0be-9fc1fc1bdafe)

Separately, the cursor changes between the canvas and UI.

![240200162-5e9b97a7-c088-4524-9715-991813a4ef8b](https://github.com/tldraw/tldraw/assets/15892272/847e4d29-6ae7-47db-88ad-5cab24d0e960)

❌ On the one hand, this can be jarring.
✅ On the other hand, perhaps the user would expect to have their *own* cursor on the UI.
ℹ️ Also, FigJam's canvas cursor is a completely different style to most OS cursors. This means that the swap feels more 'deliberate', and it's less of an uncanny-valley replacement.

![240201011-d9ffdc3b-650b-4a5f-89eb-216077b7d341](https://github.com/tldraw/tldraw/assets/15892272/e4e6a7e8-b89a-46e6-b53a-8d6128b2204f)

## Framer

- They use **browser cursors** in most places.
- They use **image cursors** for rotations.
- They use **floating cursors** for only the cursor chat bubble.

![240211329-6f86c3a9-cc5b-4946-b078-865b79b1bc59](https://github.com/tldraw/tldraw/assets/15892272/d2e76f80-7358-4e7b-bdfd-a6d9ad7d0269)

✅ The image cursors have a *little bit* of control over rotations — but they could do more if they wanted! There's also a slight flicker that they could fix if they want.

![240211874-a2d51a94-e206-421e-85dc-5e405e6f0a84](https://github.com/tldraw/tldraw/assets/15892272/e2418ff2-c9fd-44a1-88b1-66e29629b49d)

❌ The floating cursor chat bubble lags behind your actual cursor.

![240213346-363c4407-08db-419b-92d4-4a93904b0279](https://github.com/tldraw/tldraw/assets/15892272/51331259-b89c-493b-9445-0de8793ce384)

❌ It doesn't necessarily get put in the right place either.

![240214115-2b423fd9-688c-4e74-972d-c82ef7a422cc](https://github.com/tldraw/tldraw/assets/15892272/9ac1a16f-cac0-44b6-85be-2f06105c275e)

## Honourable Mention: Miro

For some tools, Miro uses a combination of **image cursor** and **floating cursor**, which works quite well.

✅ Here, the pen cursor is an **image cursor**, which feels nice and responsive.
✅ They're able to add a bigger circle around it by **floating**.

![2023-05-23 at 11 45 37 - Turquoise Worm](https://github.com/tldraw/tldraw/assets/15892272/72d84b52-7ea8-4b4d-ab9d-8022e1d30a88)

It's a bit jarring how it swaps between cursor types on the canvas though.

![2023-05-23 at 15 53 41 - Silver Dinosaur](https://github.com/tldraw/tldraw/assets/15892272/952af867-ac93-49be-ab87-b8f681f0056d)

# What should we do?

There are three separate decisions to make:
- How do we show the base cursor in the canvas?
- How do we show the cursor chat bubble?
- What do we do in the UI?

## Base Cursor: Canvas

How do we show the **base cursor** in the canvas? Four approaches to pick from:

1. Current Approach ❌ (eg: tldraw, Framer)
Mostly browser cursors, some image cursors.
I don't think we should do this, because we can't use any customisation.

2. Floating Approach ❌ 
All floating cursors.
I don't think we should do this, because of the annoying lag behind the pointer.

3. Browser Approach ❌ (eg: Excalidraw)
All browser cursors.
We shouldn't do this, because we don't get to control rotations or customisation.

4. Image Approach ✅ (eg: FigJam)
All image cursors.
I think we should do this! It lets us customise stuff, and keep it consistent.

## Chat Bubble

How do we show a chat bubble? Two approaches to pick from:

1. Swap Approach (eg: FigJam) ❌ 
Swap out the base cursor, and replace it with a floating cursor and chat bubble.
I don't think we should do this, because there's a noticeable difference in cursor responsiveness when you're chatting. I think that snappiness of the base cursor is more important in tldraw.
It also makes the cursor change size when it happens, which doesn't look good.

2. Chase Approach (eg: Framer) ✅ 
Keep the base cursor and make a floating chat bubble chase it.
I think we should do this, because it keeps the base cursor responsive. There is a lag between the cursor and bubble though - but we can lean into this with the design, and make it seem more intentional.
By using image cursors for the base cursor, we can position the chat bubble in the right place more often (compared to Framer, who uses browser cursors for the base cursor).

## UI

If we roll our own cursors, what do we do about the UI?

1. Inconsistent Approach (eg: FigJam)
Show the default browser cursors in the UI.

2. Consistent Approach
Show our custom cursor within the UI as well.

I think that we should revisit the UI question after implementing cursor chat within the canvas.
We will be able to go either route! And the decision comes with a bunch more considerations.

# What should we do?

tl;dr

- In the canvas, we should _always_ replace the cursor with custom images (like FigJam).
- Your cursor chat bubble should 'chase' the cursor (like Framer). We shouldn't swap out the base cursor.
- For now, we should keep the UI as using the default **browser cursors** (like everyone else). We should revisit later though.

**Comment by ds300** [comment]
👍🏼 wonderful breakdown, thanks @TodePond! I'm on board with your suggestions.

<img width="774" alt="image" src="https://github.com/tldraw/tldraw/assets/1242537/d1542efe-14f6-48e9-9945-8e97a04d0a78">

We use SVGs for these. Does it only rasterize them at a particular resolution?

**Comment by forresto** [comment]
Instead of janky updating, what about a tight physics animation to move the custom thing around. If tied to a doodling tool, it might be an enhancement. Similar to how perfect-handwriting smooths strokes, in the feel of the cursor itself? are you already doing something like this with [multiplayer cursors](https://mas.to/@tldraw/110231425101595406)? 

I liked your multiplayer puppetry TLDraw demo... that with physics floppy ears would be really fun.

**Comment by TodePond** [comment]
We're not doing any smoothing with cursors at the moment - the multiplayer is often smooth enough that it feels like that :) But that could help to improve it in the future!

I'm gonna close this issue now, as custom SVG cursors are now ready to release (should come out soon today).

There's gonna be more work to do around cursor chat though!!
